lsm: One radix buffer to rule them all.

toziegler opened the pull request:
Previously, each table owned its own radix sort buffer. 
This PR centralizes the buffer in `forest.zig` so it’s shared across all tables and safes a decent amount of memory.

I implemented it multiple times with different designs and I think this was the best one for multiple reasons: 

- It’s a storage-layer concern; keeping it out of the state machine preserves proper separation of responsibilities. The buffer now lives in the forest and is shared from there.
- This is the most direct implementation. It avoids passing the buffer through via parameters (including scan paths via the state machine).
- The naming reflects that it is tied to the radix sort, which makes grepping and maintainability easier, e.g., if we would change radix sort to an inplace sort etc.

toziegler committed:
lsm: share sort buffer.

toziegler committed:
lsm: add alginment check (it takes two to contract).

toziegler committed:
lsm: rename shared radix buffer.

toziegler commented:
@matklad I assigned you, since we talked about it in the original PR, but feel free to reassign if you are busy!

matklad commented on src/lsm/forest.zig:
```diff
@@ -178,6 +179,20 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
+            max_size = @max(max_size, size);
+            max_alignment = @max(max_alignment, alignment);
+        }
+        break :blk .{ max_size, max_alignment };
```
Assert that sizes is positive and that max_alignment is power of two

matklad commented on src/lsm/forest.zig:
```diff
@@ -286,12 +307,22 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
 
+            forest.radix_scratch.buffer =
+                try allocator.alignedAlloc(u8, radix_scratch_alignment, radix_scratch_size);
+            errdefer allocator.free(forest.radix_scratch.buffer);
```
This is incorrect, right, because we use wrong alignment? 

I'd maybe add `RadixBuffer.init`/`deinit`, and a field to track alignment. That way you could also assert at runtime when using the buffer that alignment is correct. Remember, asserts come in pairs, once you establish invariant, and once when you are relying on the invariant being established. 

* We compute alignment by looping over the trees. 
* But then, when using an individual tree, we assert <= to make sure we accounted for the tree when computing the alignment initially

matklad commented on src/lsm/radix_buffer.zig:
```diff
@@ -0,0 +1,4 @@
+pub const RadixBuffer = struct {
```
Let's call it `ScratchMemory`? I think that better captures it. It is both a ScratchMemory and a RadixBuffer, because it is scratch memory used as a radix buffer. But fro the implementation itself ,we care about it being scratch memory. 

And

```
radix_buffer: ScratchMemory
```

I think reads best!

matklad commented on src/lsm/forest.zig:
```diff
@@ -317,6 +348,11 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
+            // This matches the allocation alignment.
+            allocator.free(
+                @as([]align(radix_scratch_alignment) u8, @alignCast(forest.radix_scratch.buffer)),
+            );
```
Assert state of `in_use`

matklad commented on src/lsm/forest.zig:
```diff
@@ -346,6 +382,7 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
                 .compaction_schedule = forest.compaction_schedule,
 
                 .scan_buffer_pool = forest.scan_buffer_pool,
+                .radix_scratch = forest.radix_scratch,
```
ditto, assert state of `in_use`

matklad commented on src/lsm/groove.zig:
```diff
@@ -608,6 +610,7 @@ pub fn GrooveType(
+            radix_scratch: *RadixBuffer,
             options: Options,
         ) !void {
             assert(options.tree_options_object.batch_value_count_limit *
```
assert in_used

matklad commented on src/lsm/tree.zig:
```diff
@@ -103,6 +104,7 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type) type {
+            radix_scratch: *RadixBuffer,
             config: Config,
             options: Options,
         ) !void {
```
assert `in_use`

matklad commented on src/lsm/table_memory.zig:
```diff
@@ -303,6 +306,7 @@ pub fn TableMemoryType(comptime Table: type) type {
             table: *TableMemory,
             allocator: mem.Allocator,
             mutability: std.meta.Tag(Mutability),
+            radix_scratch: *RadixBuffer,
```
The order feels wrong here. `radix_scratch` is a resource used by the implementation. `mutability` and `name` are parameters configuring the implementation. 

So, `radix_buffer, mutability, name` or `mutability, name, radix_buffer` make sense, but the middle one doesn't. I think radix_buffer right after the allocator makes most sense.

matklad commented on src/lsm/table_memory.zig:
```diff
@@ -328,31 +332,17 @@ pub fn TableMemoryType(comptime Table: type) type {
-            if (table.mutability == .mutable) {
-                allocator.free(table.mutability.mutable.values_scratch);
-            }
             allocator.free(table.values);
```
assert in_use

matklad commented on src/lsm/table_memory.zig:
```diff
@@ -328,31 +332,17 @@ pub fn TableMemoryType(comptime Table: type) type {
                 .immutable => .{ .immutable = .{} },
                 .mutable => .{ .mutable = .{
-                    .values_scratch = table.mutability.mutable.values_scratch,
+                    .radix_scratch = table.mutability.mutable.radix_scratch,
```
assert in_use

matklad commented on src/lsm/table_memory.zig:
```diff
@@ -585,9 +575,25 @@ pub fn TableMemoryType(comptime Table: type) type {
+                table.mutability.mutable.radix_scratch.buffer[0..radix_scratch_size],
+            );
+            assert(@as(usize, @intFromPtr(radix_scratch_values.ptr)) % @alignOf(Value) == 0);
+            assert(radix_scratch_values.len == table.values.len);
```
This smells like a method on Scratch memory: 

```zig
const scratch_values = table.mutability.mutable.radix_buffer.acquire(Value, table.count());
defer table.mutability.mutable.radix_buffer.release(scratch_values);
```

(we don't have to pass the slice on release, but that allows for extra asserts, and also hits the reader about the appropriate protocol when using ScratchMemory API)

matklad commented on src/lsm/table_memory.zig:
```diff
@@ -665,11 +671,13 @@ const TestHelper = struct {
         comptime TableType: type,
         gpa: std.mem.Allocator,
         value_count_limit: u32,
+        radix_scratch: *RadixBuffer,
```
Kinda urgh that we need to pass the buffer in even for immutable table, though I think `make-illegal-states-unrepresentable`ing that is not worth the hassle!

matklad commented on src/lsm/table_memory.zig:
```diff
@@ -585,9 +575,25 @@ pub fn TableMemoryType(comptime Table: type) type {
+            defer {
+                table.mutability.mutable.radix_scratch.in_use = false;
+                assert(!table.mutability.mutable.radix_scratch.in_use);
+            }
```
Love how, because the API is fully synchronous, we don't _really_ have to worry about buffer re-use here, we statically know that there's no concurrency while the buffer is in use!

matklad commented on src/lsm/table_memory.zig:
```diff
@@ -703,17 +713,29 @@ test "table_memory: merge and absorb (last wins across streams)" {
+        ),
+        .in_use = false,
+    };
+    defer alloc.free(@as([]align(@alignOf(Value)) u8, @alignCast(radix_scratch.buffer)));
```
Again, init/deinit pair would simplify this.

matklad commented on src/lsm/table_memory.zig:
```diff
@@ -703,17 +713,29 @@ test "table_memory: merge and absorb (last wins across streams)" {
 
     const alloc = testing.allocator;
 
+    var radix_scratch = RadixBuffer{
```
```suggestion
    var radix_scratch: RadixBuffer = .{
```

is a touch more idiomatic (though, we'll change that to `: RadixBuffer = try .init`) anyways!

matklad commented on src/lsm/table_memory.zig:
```diff
@@ -804,17 +838,29 @@ test "table_memory (secondary): annhiliation yields zero after deduplicate" {
+        ),
+        .in_use = false,
+    };
+    defer alloc.free(@as([]align(@alignOf(Value)) u8, @alignCast(radix_scratch.buffer)));
```
This is a very specific code smell! These lines are related three times in tests, but only once in production code. This is baad, because it makes refactoring disproportionally painful. For example, to introduce init/deinit, you need to change prod code only once, but there are there test places to fix. 

So, it's better to proactively extract such common setup-code into a fixture, something like `TestHelper.create_scratch_buffer` (see also `fixtures.zig`). Though, if you refactor that to init/deinit, it'll be only two lines, so maybe not worth extracting!

matklad commented on src/lsm/radix_buffer.zig:
```diff
@@ -0,0 +1,4 @@
+pub const RadixBuffer = struct {
+    buffer: []u8,
+    in_use: bool,
+};
```
Good call to extract it into a file, but:

* Add `//!` explaining the context around the data structure
* Move relevant logic to this file (init/deinit, and acquire/release of typed slices)
* It would be cool to somehow move the pattern of computing the max size&alignment here as well, but I don't see a good way to do that. 

Also consider using file-as-astruct pattern in zig:

```
//! ScratchMemory is bla-bla-bla
buffer: []u8, 
alignment: u8, 
busy: bool, 

const ScratchMemory = @This();

pub fn init(gpa: Allocator, size: usize, alignment: u8) !ScratchMemory
pub fn deinit(scratch: *ScratchMemory, gpa: Allocator) void

pub fn acquire(T: type, count: usize) []T
pub fn release(T: type, slice: []T) void
```

matklad commented on src/lsm/radix_buffer.zig:
```diff
@@ -0,0 +1,4 @@
+pub const RadixBuffer = struct {
+    buffer: []u8,
+    in_use: bool,
```
`in_use` is a bit of an awkward name, with two words. I grepped the code base for `: bool` to find if we already have a good name somewhere, and `busy: bool` was one good option.

batiati replied:
The CDC runner has a similar buffer, and we use the pattern of an internal variable `state: enum { free, busy }` and two functions `acquire() []u8` and `release() void` that assert the correct usage.

matklad replied:
Ah yeah! I do have "`enum` is almost always better than `bool`" rule of thumb, but always forget using it!

matklad commented on src/lsm/table_memory.zig:
```diff
@@ -588,6 +588,7 @@ pub fn TableMemoryType(comptime Table: type) type {
                 Value,
                 table.mutability.mutable.radix_scratch.buffer[0..sort_buffer_size],
             );
+            assert(@as(usize, @intFromPtr(sort_buffer_values.ptr)) % @alignOf(Value) == 0);
```
Ah, ok, I actually missed this, yeah, that' the check I had in mind, gut!

batiati replied:
use `std.mem.isAligned(@intFromPtr(sort_buffer_values.ptr), @alignOf(Value))` instead

batiati commented on src/lsm/table_memory.zig:
```diff
@@ -585,9 +575,25 @@ pub fn TableMemoryType(comptime Table: type) type {
+            table.mutability.mutable.radix_scratch.in_use = true;
+            defer {
+                table.mutability.mutable.radix_scratch.in_use = false;
+                assert(!table.mutability.mutable.radix_scratch.in_use);
```
Assert before?

```suggestion
                assert(table.mutability.mutable.radix_scratch.in_use);
                table.mutability.mutable.radix_scratch.in_use = false;
```

toziegler committed:
lsm: refactor and rename `radix_scratch.zig` to `scratch_memory.zig`.

toziegler committed:
lsm: reorder `init` params in `table_memory.zig`.

toziegler committed:
lsm: use idiomatic init pattern for `ScratchMemory`.

toziegler committed:
lsm: assert that ScratchMemory is `.free` in all places.

toziegler committed:
lsm: fix `ScratchMemory` sizing in `tree_fuzz.zig`.

toziegler committed:
lsm: add documentation for `ScratchMemory`.

toziegler committed:
lsm: assert that size is greater than zero for `ScratchMemory`.

toziegler committed:
lsm: rename `radix_scratch` to `radix_buffer`.

toziegler commented:
Rebased.

toziegler commented:
Just to document @matklad 's   idea here as well: `ScratchMemory` is aligned to the system page size. This guarantees the buffer is page-aligned, so it can be safely cast to types whose alignment requirement is ≤ the page size (see [alignment](https://ziglang.org/documentation/0.14.0/#Alignment)) (which we ensure now in an `assert`). A nice side-effect is that we do not need to calculate the `max_alignment` of all involved types.

matklad commented on src/lsm/scratch_memory.zig:
```diff
@@ -0,0 +1,80 @@
+
+pub fn release(scratch: *ScratchMemory, T: type, slice: []T) void {
+    assert(scratch.state == .busy);
+    assert(std.mem.isAligned(@intFromPtr(slice.ptr), @alignOf(T)));
```
```
comptime assert(@alignOf(T) < std.heap.page_size_min);
```

The trick is that we want to catch this at compile time, and we can, because we comptime know both the type (T), and the alignment we used for allocation (page_size_min).

It's still good to do a runtime assert on the actual pointer as a defense in depth though!

matklad replied:
(I didn't realize that I am looking at the release, not acquire. Though acquire also misses comptime (I _think_ it is effectively comptime, given Zig's evaluation rules, but its worth to make it syntactically obvious))

matklad replied:
Oh, in release, we can actually assert tighter:

```
assert(@intFromPtr(slice.ptr) == @intFromBtr(scratch.buffer.ptr));
assert(slice.len * @sizeOf(T) <= scratch.buffer.len);
```

matklad replied:
>I think it is effectively comptime, given Zig's evaluation rules, but its worth to make it syntactically obvious)

For the reference, this was wrong! In

```
assert(@alignOf(T) <= std.heap.page_size_min);
```

The _argumetn_ to `assert` is comptime  known, because `@alignOf` is a builtin, and `std.heap.page_size_min` is a const. But the `assert` call itself is still a call to a non-inline function. SO, while it is partially evaluated to `assert(true)`, the assert call itself is runtime, so `comptime` is actually necessary!

toziegler replied:
Nice find! Semantically having `comptime` is also nicer for the reader <3.

batiati commented on src/lsm/scratch_memory.zig:
```diff
@@ -0,0 +1,80 @@
+//! ScratchMemory is a page-aligned scratch buffer meant for situations where a buffer is required
+//! (e.g., radix sort) and can be shared between components.
+//! The buffer is page-aligned so that smaller alignments are trivially satisfied. See:
+//! https://ziglang.org/documentation/0.7.0/#Alignment
```
Tobi the archeologist 🤣 

```suggestion
//! https://ziglang.org/documentation/0.14.1/#Alignment
```

toziegler replied:
Wow,  how did that happen :-D!

matklad commented on src/lsm/forest.zig:
```diff
@@ -286,12 +290,32 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
             };
 
+            const radix_buffer_size: usize = comptime blk: {
+                var max_size: usize = 0;
```
size_max, to align with 
size

matklad commented on src/lsm/scratch_memory.zig:
```diff
@@ -0,0 +1,80 @@
+}
+
+pub fn deinit(scratch: *ScratchMemory, gpa: std.mem.Allocator) void {
+    assert(scratch.state == .free);
```
⚡

toziegler committed:
lsm: Update zig version in comment.

Co-authored-by: Rafael Batiati <rbatiati@gmail.com>

toziegler committed:
lsm: `size_max` instead of `max_size` per TigerStyle.

toziegler committed:
lsm: comptime alignment check for `ScratchMemory`.

matklad commented on src/lsm/scratch_memory.zig:
```diff
@@ -0,0 +1,80 @@
+
+pub fn deinit(scratch: *ScratchMemory, gpa: std.mem.Allocator) void {
+    assert(scratch.state == .free);
+    gpa.free(scratch.buffer);
```
```suggestion
    gpa.free(scratch.buffer);
    scratch.* = undefined;
```

matklad commented on src/lsm/scratch_memory.zig:
```diff
@@ -0,0 +1,80 @@
+    assert(scratch.state == .free);
+    assert(count * @sizeOf(T) <= scratch.buffer.len);
+    // A pointer with a larger alignment can be cast into one with a smaller alignment.
+    assert(@alignOf(T) <= std.heap.page_size_min);
```
```suggestion
    comptime assert(@alignOf(T) <= std.heap.page_size_min);
```

matklad commented on src/lsm/scratch_memory.zig:
```diff
@@ -0,0 +1,80 @@
+    scratch.state = .free;
+}
+
+test "ScratchMemory basic" {
```
❤️ 

could also just `test ScratchMemory {`, test name can be either a string, or another declaration (so, an arbitrary identifier won't compile, only an existing one will).

toziegler committed:
lsm: set scratch to `undefined` after free.

Co-authored-by: Alex Kladov <aleksey.kladov@gmail.com>

matklad reviewed (dismissed):
In release, could add assert that slice.ptr is the same as buffer.ptr, but that's not _really_ needed, LGTM as is.

toziegler committed:
lsm: add another assertion in `ScratchMemory`.

toziegler commented:
> In release, could add assert that slice.ptr is the same as buffer.ptr, but that's not _really_ needed, LGTM as is.

I added the assertion - better safe than sorry, so might need another quick approve :-).  EDIT: Also rebased.
